### PR TITLE
Stream zip or tar

### PR DESCRIPTION
**Pull request by VicDeo**
download directory content as tar for macs.
~~Depends on https://github.com/owncloud/3rdparty/pull/197~~
https://github.com/owncloud/3rdparty/pull/198
Ref https://github.com/owncloud/core/issues/17930

P.S. [Elementary tar structures spec](https://gist.github.com/VicDeo/464a76021a6fae10bf4c) for okteta hex editor

### CODE REVIEW

File: lib/private/streamer.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+use \ZipStreamer\ZipStreamer;
+use DeepDiver1975\TarStreamer\TarStreamer;
+
+class OC_Streamer {
```
Comment (DeepDiver1975) [inline lib/private/streamer.php]: \OC\Streamer

**Comment by DeepDiver1975** [comment]
zips and tars are generated and the tar can be opened on my linux machine.

Test on Mac os is still pending @LukasReschke

File: lib/private/streamer.php [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+class Streamer {
+	// array of regexp. Matching user agents will get tar instead of zip
+	private $preferTarFor = [ '/macintosh|mac os x/i' ];
+
+	// streamer instance
+	private $streamerInstance;
+	
+	/** @var string*/
+	private $extension;
```
Comment (DeepDiver1975) [inline lib/private/streamer.php]: unused

File: lib/private/streamer.php [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+		if ($request->isUserAgent($this->preferTarFor)) {
+			$this->streamerInstance = new TarStreamer();
+		} else {
+			$this->streamerInstance = new ZipStreamer();
+		}
+	}
+	
+	/**
+	 * Send HTTP headers
+	 * @param string name
```
Comment (DeepDiver1975) [inline lib/private/streamer.php]: $name

File: lib/private/streamer.php [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	}
+	
+	/**
+	 * Send HTTP headers
+	 * @param string name 
+	 * @return bool
+	 */
+	public function sendHeaders($name){
+		$extension = $this->streamerInstance instanceof ZipStreamer ? '.zip' : '.tar';
+		return $this->streamerInstance->sendHeaders($name . $extension);
```
Comment (DeepDiver1975) [inline lib/private/streamer.php]: sendHeaders implementations don't return anything

File: lib/private/streamer.php [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+	 * @param string $dir
+	 * @param string $internalDir
+	 */
+	public function addDirRecoursive($dir, $internalDir='') {
+		$dirname = basename($dir);
+		$rootDir = $internalDir . $dirname;
+		if (!empty($rootDir)) {
+			$this->streamerInstance->addEmptyDir($rootDir);
+		}
+		$internalDir.= $dirname .= '/';
```
Comment (DeepDiver1975) [inline lib/private/streamer.php]: double .= - looks odd

File: lib/private/streamer.php [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+		$extension = $this->streamerInstance instanceof ZipStreamer ? '.zip' : '.tar';
+		return $this->streamerInstance->sendHeaders($name . $extension);
+	}
+	
+	/**
+	 * Stream directory recursively
+	 * @param string $dir
+	 * @param string $internalDir
+	 */
+	public function addDirRecoursive($dir, $internalDir='') {
```
Comment (DeepDiver1975) [inline lib/private/streamer.php]: typo

**Comment by DeepDiver1975** [comment]
THX @VicDeo

**Comment by LukasReschke** [comment]
Bugs:

- [ ] Create a folder named `test">`, download it. Will be named `test--.tar-;`

Functionality:
- [x] Downloading an empty folder and extracting work on OS X
- [x] Downloading an folder with some files in it and extracting work on OS X
- [x] Downloading an folder with some files and subfolders in it and extracting work on OS X
- [x] Downloading an folder with only subfolders in it and extracting work on OS X
- [x] Downloading multiple folders works on OS X

**Comment by DeepDiver1975** [comment]
- [ ] tar file with a size of 4gb+

**Comment by DeepDiver1975** [comment]
> Create a folder named test">, download it. Will be named test--.tar-;

same for zip?

**Comment by LukasReschke** [comment]
> same for zip?

Nope. ZIP on master produces `test2--.zip`

**Comment by LukasReschke** [comment]
>  tar file with a size of 4gb+

TAR files totalling more than 5GB seem to work fine here on OS X.

**Comment by LukasReschke** [comment]
> same for zip?

Bug in the library :speak_no_evil: https://github.com/owncloud/3rdparty/pull/197/files#r40337737

**Comment by VicDeo** [comment]
>Create a folder named test">, download it. Will be named test--.tar-;
@LukasReschke 
now it is  `test--.tar`. Is it ok? :)

**Comment by DeepDiver1975** [comment]
> TAR files totalling more than 5GB seem to work fine here on OS X.

nice :rocket:

**Comment by DeepDiver1975** [comment]
TarStreamer now lives in the ownCloud orga - this is ready to be finally tested and merged

@VicDeo @PVince81 @LukasReschke @rperezb @SergioBertolinSG THX

**Comment by MorrisJobke** [comment]
I tested it also and all still works here on with unzip (on linux). :+1: from me, if this solves the OS X issue.

**Comment by DeepDiver1975** [comment]
:+1: let's move this in

**Comment by DeepDiver1975** [comment]
@LukasReschke final objections? THX

**Comment by LukasReschke** [comment]
:+1: 

Merge at will.

**Comment by DeepDiver1975** [comment]
3rdparty merged ... updating submodule ... soooo boring

**Comment by karlitschek** [comment]
@DeepDiver1975 @VicDeo We have to backport this at least to 8.1, better also 8.0

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 @VicDeo We have to backport this at least to 8.1, better also 8.0

What is the time frame on this backport? I'd prefer some quarantine before another bug epidemic on this topic hits us ....

**Comment by karlitschek** [comment]
@MTRichards @cmonteroluque @msrex Opinions about the timing?

**Comment by msrex** [comment]
does it work on our infrastructure? if yes, we can continue to quarantine there for a few days...

**Comment by DeepDiver1975** [comment]
> does it work on our infrastructure? if yes, we can continue to quarantine there for a few days...

If we want to dog food this feature we would need to backport it first and deploy a patch version on our server - do we want that?

**Comment by karlitschek** [comment]
Sounds like a good idea. Let's do a backport PR and deploy the patch on our machine. We then merge it for 8.1.4 which is a few weeks away anyways. Does this work for everyone?

**Comment by MorrisJobke** [comment]
> Sounds like a good idea. Let's do a backport PR and deploy the patch on our machine. We then merge it for 8.1.4 which is a few weeks away anyways. Does this work for everyone?

Sounds like a plan

**Comment by DeepDiver1975** [comment]
I'll prepare the backport tonight

**Comment by cmonteroluque** [comment]
@DeepDiver1975 thanks!

**Comment by rperezb** [comment]
@SergioBertolinSG please

**Comment by rperezb** [comment]
Suggested test plan:

## 8.2

Expected result in every case: archives can be opened without error using the os archive tools

Test          | Result: MAC  | Result: Windows | Result: Linux  | Result: Mobile
------------- | -------------| ------------- | ------------- | -------------
File is downloaded  | | | |
Folder is downloaded  | | ||
Select some files within a folder and download them  | | | |

**Comment by DeepDiver1975** [comment]
And also add test cases with files and folder with a name longer then 100 chars - see https://github.com/owncloud/TarStreamer/issues/4#issuecomment-146319813

**Comment by SergioBertolinSG** [comment]
One file downloaded in a public share in OS X downloads a tar file instead of the file directly. Is this a problem? I find it annoying because it is just one file and you have to open it twice (open .tar + real open).

**Comment by DeepDiver1975** [comment]
> One file downloaded in a public share in OS X downloads a tar file instead of the file directly. Is this a problem? I find it annoying because it is just one file and you have to open it twice (open .tar + real open).

This was fixed yesterday - please use current master for testing

**Comment by SergioBertolinSG** [comment]
Ok, I was using 8.2RC1.

**Comment by SergioBertolinSG** [comment]
When downloading a public shared folder the name of the file is always download.tar, instead of the folder name, inside the folder has the proper name. 
When selecting several files of this shared folder, the downloaded file is called like the   shared folder.
It should be the opposite isn't? the selected files downloaded in download.tar and the whole folder with {folder_shared_name}.tar

**Comment by PVince81** [comment]
@SergioBertolinSG maybe, can you raise a ticket for that ?

**Comment by DeepDiver1975** [comment]
@nickvergessen another testcase for smashbox?

**Comment by nickvergessen** [comment]
> @nickvergessen another testcase for smashbox?

Once the issue is fixed I can try to get ti working. Currently I specify the full-path in the curl request that grabs the zip. So that would need a check for the proposed file name instead.

**Comment by SergioBertolinSG** [comment]
@PVince81 raised https://github.com/owncloud/core/issues/19646

**Comment by DeepDiver1975** [comment]
I missinterpreted the original report - the behavior is like that since forever.
@SergioBertolinSG I'm sorry

**Comment by DeepDiver1975** [comment]
> And also add test cases with files and folder with a name longer then 100 chars - see owncloud/TarStreamer#4 (comment)

@SergioBertolinSG can you test this on macos please? THX

**Comment by SergioBertolinSG** [comment]
@DeepDiver1975 Yes I did, works fine. Same in windows and linux.

**Pull request merged**